libs/@blockprotocol/graph/src/codegen/initialize/metadata/generate-metadata-schema.ts

```diff
@@ -0,0 +1,246 @@
+          dataTypeId: { const: dataTypeSchema.$id },
+        },
+        required: ["dataTypeId"],
+      },
+    },
+    required: ["metadata", "value"],
+  };
+};
+
+type EntityParentIdentifiers = {
+  title: string;
+  $id: VersionedUrl;
+};
+
+type ObjectWithMetadataParams = {
+  properties: EntityType["properties"];
+  required: string[];
+  entityParentIdentifiers: EntityParentIdentifiers | null;
+};
+
+export function generatePropertiesObjectWithMetadataSchema({
+  properties,
+  required,
+  entityParentIdentifiers,
+}: ObjectWithMetadataParams & {
+  entityParentIdentifiers: null;
+}): PartialJsonSchema;
+export function generatePropertiesObjectWithMetadataSchema({
+  properties,
+  required,
+  entityParentIdentifiers,
+}: ObjectWithMetadataParams & {
+  entityParentIdentifiers: EntityParentIdentifiers;
+}): JsonSchema;
+/**
+ * Generate a schema for a properties object with metadata.
+ *
+ * If entityParentIdentifiers is provided, it will also assign an $id to the type so that it appears separately.
+ * The $id is derived from the parent and assumes that the parent is creating only a single properties object
+ * – do not use this for properties objects defined within a property type, which may include multiple properties objects.
+ */
+export function generatePropertiesObjectWithMetadataSchema({
+  properties,
+  required,
+  entityParentIdentifiers,
+}: ObjectWithMetadataParams): JsonSchema | PartialJsonSchema {
+  const title = entityParentIdentifiers
+    ? generateMetadataSchemaTitles({
+        title: `${entityParentIdentifiers.title} Properties`,
+      }).valueWithMetadataTitle
+    : undefined;
+
+  const $id = entityParentIdentifiers
+    ? (generateMetadataSchemaIdentifiers({
+        $id: entityParentIdentifiers.$id,
+      }).valueWithMetadata$id.replace(
+        "/with-metadata/",
+        "/properties-with-metadata/",
+      ) as VersionedUrl)
+    : undefined;
+
+  const propertiesSchema = Object.entries(properties).reduce<PartialJsonSchema>(
+    (acc, [baseUrl, refSchema]) => {
+      const propertyWithMetadataRef = generateMetadataSchemaIdentifiers({
+        $id: "items" in refSchema ? refSchema.items.$ref : refSchema.$ref,
+      }).valueWithMetadata$id;
+
+      if ("items" in refSchema) {
+        acc[baseUrl] = {
+          type: "object",
+          properties: {
+            value: {
+              type: "array",
+              items: {
+                $ref: propertyWithMetadataRef,
+              },
+            },
+            metadata: arrayMetadataSchema,
+          },
+          required: ["value"],
+        };
+      } else {
+        acc[baseUrl] = {
+          $ref: propertyWithMetadataRef,
+        };
+      }
+
+      return acc;
+    },
+    {},
+  );
+
+  return {
+    type: "object",
+    title,
+    $id,
+    kind,
+    properties: {
+      metadata: objectMetadataSchema,
+      value: {
+        type: "object",
+        properties: propertiesSchema,
+        required,
+      },
+    },
```

Comment:
Is this the location you were wondering if we want `value` in the first place? So if we want `value` at the top level?

Reply:
Yes, because I'm not sure what it means when applied to the whole entity (as compared to a nested object, where it could be taken to mean 'existince of this thing'). But it's not a big deal, fine to be consistent to start